snapcraft/internal/repo/_deb.py

```diff
@@ -411,12 +458,17 @@ def _format_sources_list(sources_list, *,
         suffix = 'ubuntu-ports'
         security = 'ports'
 
-    return string.Template(sources_list).substitute({
-        'prefix': prefix,
-        'release': release,
-        'suffix': suffix,
-        'security': security,
-    })
+    try:
+        return string.Template(sources_list).substitute({
+            'prefix': prefix,
+            'release': release,
+            'suffix': suffix,
+            'security': security,
+            'arch': ' [arch={}]'.format(deb_arch) if foreign else '',
+        })
+    except KeyError as e:
+        raise errors.SnapcraftError(
+            'Sources list is missing the variable ${' + str(e) + '}')
```

Comment:
I fell prey to this as well: this error message is a little backward: a `KeyError` is raised if the template is not satisfied, not if a substitution was requested but not performed. Code will be more clear:

```
>>> import string
>>> t = '$food is delicious'
>>> string.Template(t).substitute(food='sushi')
'sushi is delicious'
>>> string.Template(t).substitute(food='sushi', bar='baz')
'sushi is delicious'
>>> string.Template(t).substitute(bar='baz')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.5/string.py", line 129, in substitute
    return self.pattern.sub(convert, self.template)
  File "/usr/lib/python3.5/string.py", line 119, in convert
    val = mapping[named]
KeyError: 'food'
```

So the error should probably be something more along the lines of "Unable to satisfy template in sources list: $foo".

Reply:
There doesn't seem to be a way to warn in the reverse situation... hmm...

Reply:
You made me wonder what's actually an error here: if you're using a varaible ${foo} that we don't know how to fill in, that seems wrong - but using all the variables isn't necessary.

So, I updated the error based on your wording suggestion and added two test cases for the above mentioned cases, one of them failing, the other one passing. It seems to make sense this way... but let me know what you think.